File: 04-april/01-robot-collisions/solution.js
var survivedRobotsHealths = function(positions, healths, directions) {
    const n = positions.length;
    const robots = Array.from({length: n}, (_, i) => i).sort((a, b) => positions[a] - positions[b]);
    const h = [...healths];
    const stack = [];

    for (const i of robots) {
        if (directions[i] === 'R') {
            stack.push(i);
        } else {
            while (stack.length && h[i] > 0) {
                const top = stack[stack.length - 1];
                if (h[top] > h[i]) {
                    h[top]--;
                    h[i] = 0;
                } else if (h[top] < h[i]) {
                    h[i]--;
                    h[top] = 0;
                    stack.pop();
                } else {
                    h[top] = 0;
                    h[i] = 0;
                    stack.pop();
                }
            }
        }
    }

    return h.filter(x => x > 0);
};

File: 04-april/01-robot-collisions/solution.java
class Solution {
    public List<Integer> survivedRobotsHealths(int[] positions, int[] healths, String directions) {
        int n = positions.length;
        Integer[] robots = new Integer[n];
        for (int i = 0; i < n; i++) robots[i] = i;
        java.util.Arrays.sort(robots, (a, b) -> positions[a] - positions[b]);

        Deque<Integer> stack = new ArrayDeque<>();

        for (int i : robots) {
            if (directions.charAt(i) == 'R') {
                stack.push(i);
            } else {
                while (!stack.isEmpty() && healths[i] > 0) {
                    int top = stack.peek();
                    if (healths[top] > healths[i]) {
                        healths[top]--;
                        healths[i] = 0;
                    } else if (healths[top] < healths[i]) {
                        healths[i]--;
                        healths[top] = 0;
                        stack.pop();
                    } else {
                        healths[top] = 0;
                        healths[i] = 0;
                        stack.pop();
                    }
                }
            }
        }

        List<Integer> result = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            if (healths[i] > 0) result.add(healths[i]);
        }
        return result;
    }
}
